### PR TITLE
Rework `non_local_definitions` lint to only use a syntactic heuristic

### LINKED ISSUE #126768: `non_local_definitions` lint fires for `impl Trait for NonLocalType<SomeLocalType>`, probably shouldn't
#120363 introduces some warnings that seem like bad advice. Nightly now warns if a function has an internal implementation of a trait on a non-internal (to that function) type. That seems like a reasonable warning in the general case, but it doesn't seem so reasonable for types like `Box<Inner>`, where `Inner` is defined within the function and cannot be named outside it.

It rather seems like something like the orphan rule should apply. You should be able to impl traits on `Foo<T>` as long as `T` is defined at the same level as the `impl`.

[Playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=a78ede63d0901a5fd6d8f37c7df6aaf2)

I tried this code on nightly:

```rust
#![allow(dead_code)]
trait Tr {}
fn foo() {
    struct Bar;
    impl Tr for Box<Bar> {}
}
```

I expected to see this happen: no errors/warnings

Instead, this happened:

```
warning: non-local `impl` definition, `impl` blocks should be written at the same level as their item
 --> src/lib.rs:5:5
  |
5 |     impl Tr for Box<Bar> {}
  |     ^^^^^--^^^^^---^^^^^
  |          |      |
  |          |      `Box` is not local
  |          `Tr` is not local
  |
  = note: `impl` may be usable in bounds, etc. from outside the expression, which might e.g. make something constructible that previously wasn't, because it's still on a publicly-visible type
  = note: an `impl` is never scoped, even when it is nested inside an item, as it may impact type checking outside of that item, which can be the case if neither the trait or the self type are at the same nesting level as the `impl`
help: move the `impl` block outside of this function `foo`
 --> src/lib.rs:3:1
  |
3 | fn foo() {
  | ^^^^^^^^
4 |     struct Bar;
  |     ---------- may need to be moved as well
  = note: this lint may become deny-by-default in the edition 2024 and higher, see the tracking issue <https://github.com/rust-lang/rust/issues/120363>
  = note: `#[warn(non_local_definitions)]` on by default
```

### Meta

`rustc --version --verbose` (well, from the playground):
```
Build using the Nightly version: 1.81.0-nightly
(2024-06-19 d8a38b00024cd7156dea)
```

<details><summary>Backtrace</summary>
<p>

```
<backtrace>
```

</p>
</details>

### PR DESCRIPTION
**Pull request by Urgau**
This PR reworks the `non_local_definitions` lint to only use a syntactic heuristic, i.e. not use a type-system logic for whenever an `impl` is local or not.

Instead the new logic wanted by T-lang in https://github.com/rust-lang/rust/issues/126768#issuecomment-2192634762, which is to consider every paths in `Self` and `Trait` and to no longer use the type-system inference trick.

@rustbot labels +L-non_local_definitions
Fixes #126768

### CODE REVIEW

File: tests/ui/lint/non-local-defs/exhaustive.rs [hunk 1]
```diff
@@ -57,18 +55,13 @@ fn main() {
...
     impl Trait for *mut [InsideMain] {}
-    //~^ WARN non-local `impl` definition
     impl Trait for [InsideMain; 8] {}
-    //~^ WARN non-local `impl` definition
     impl Trait for (InsideMain,) {}
-    //~^ WARN non-local `impl` definition
     impl Trait for fn(InsideMain) -> () {}
-    //~^ WARN non-local `impl` definition
     impl Trait for fn() -> InsideMain {}
-    //~^ WARN non-local `impl` definition
```
Comment (Urgau) [inline tests/ui/lint/non-local-defs/exhaustive.rs:71]: @traviscross can you confirm that those changes here and in other tests are indeed what T-lang wants?

File: tests/ui/lint/non-local-defs/generics.rs [hunk 2]
```diff
@@ -1,8 +1,6 @@
 //@ check-pass
 //@ edition:2021
 
-#![warn(non_local_definitions)]
-
 trait Global {}
 
 fn main() {
```
Comment (traviscross) [inline tests/ui/lint/non-local-defs/generics.rs:6]: The case that follows below here is interesting, in that it's an example of something that someone could reasonably want to do that also doesn't *feel* like a "sneaky inner impl", and where (other than allowing the lint) we wouldn't be leaving someone a good option other than to promote the trait to a wider scope.

```rust
trait Global {}
fn main() {
    trait Local {}
    impl<T: Local> Global for Vec<T> { }
    //~^ WARN non-local `impl` definition
}
```

(This is just an observation.  This clearly has non-local effects in terms of coherence checking.)

**Review by traviscross — APPROVED** [review]
I looked over the tests; looks right to me.

File: tests/ui/lint/non-local-defs/exhaustive.rs (hunk 1 above)
Comment (traviscross) [reply tests/ui/lint/non-local-defs/exhaustive.rs:71]: The changes to the tests look consistent with the decision to me.

File: compiler/rustc_lint/src/non_local_def.rs [hunk 3]
```diff
@@ -143,42 +136,28 @@ impl<'tcx> LateLintPass<'tcx> for NonLocalDefinitions {
...
-                    .unwrap_or(false);
+                // 1.5. Remove any path that doesn't resolve to a `DefId` or if it resolve to a
+                // type-param (e.g. `T`).
+                collector.paths.retain(
+                    |p| matches!(p.res, Res::Def(def_kind, _) if def_kind != DefKind::TyParam),
+                );
 
-                if impl_has_enough_non_local_candidates {
+                // 2. We check if any of path reference a "local" parent and if that the case
+                // we bail out as asked by T-lang, even trough this isn't correct from a
```
Comment (workingjubilee) [inline compiler/rustc_lint/src/non_local_def.rs:154]:
```suggestion
                // we bail out as asked by T-lang, even though this isn't correct from a
```

File: compiler/rustc_lint/src/non_local_def.rs [hunk 4]
```diff
@@ -200,76 +179,28 @@ impl<'tcx> LateLintPass<'tcx> for NonLocalDefinitions {
...
-                let (self_ty_span, self_ty_str) =
-                    self_ty_kind_for_diagnostic(&impl_.self_ty, cx.tcx);
-
-                ms.push_span_label(
-                    self_ty_span,
-                    fluent::lint_non_local_definitions_self_ty_not_local,
-                );
-                let of_trait_str = if let Some(of_trait) = &impl_.of_trait {
+                for path in &collector.paths {
+                    // FIXME: While a translatable diagnostic message can have a argument
```
Comment (workingjubilee) [inline compiler/rustc_lint/src/non_local_def.rs:186]:
```suggestion
                    // FIXME: While a translatable diagnostic message can have an argument
```

**Review by workingjubilee — COMMENTED** [review]
just typo nits

**Review by WaffleLapkin — COMMENTED** [review]
I don't know how to review this. T-lang's proposal looks to me like "let's make the lint incorrect". And I don't see why the lint is useful, if it's incorrect...

**Comment by workingjubilee** [comment]
r? lang

**Comment by workingjubilee** [comment]
cc @rust-lang/lang

"Edicts from on high" by T-lang... by which I mean, "we want this, not that, so do it"... are okay when it's clarifying something very small (sometimes someone just needs to make _a decision_ and people don't feel they have authority to make that decision). However, when something is as fundamental as the raison d'être for a lint's very existence and the implementation of it raises fundamental questions like "wait, why?", then a larger explanation is in order, unfortunately.

In other words, it's no longer clear what the goals are of this lint. I realize this is just asking for someone to basically just repeat the already-accepted RFC but with the current information we have from implementation woes and experimentation. However, this is not a waste of time: it is worth reevaluating if everything makes sense with [RFC 3347](https://rust-lang.github.io/rfcs/3373-avoid-nonlocal-definitions-in-fns.html) + [this recent comment](https://github.com/rust-lang/rust/issues/126768#issuecomment-2192634762) in mind.

**Comment by traviscross** [comment]
Replied with further context and background over in https://github.com/rust-lang/rust/issues/126768#issuecomment-2223969646.

**Comment by workingjubilee** [comment]
:thought_balloon: ...if I asked for lang to review this, and then TC nominated me, does that mean I'm on lang? Is this my punishment for ribbing T-lang one too many times? Am I now the target of my own japes? ...oh no, I'm barely even awake at 9, how am I going to make a meeting at **8**?

**Comment by traviscross** [comment]
Maybe all of that? /s.... but no, seriously that was me just handing back off after answering the lang questions.  Looking again, though, I mean to hand it back to compiler, so...

r? compiler

**Comment by michaelwoerister** [comment]
r? diagnostics

**Comment by oli-obk** [comment]
I've tried to review this every day of the past 3 days and each time finished with a bit more knowledge but mostly feeling unhappy with my understanding of it and the exact things we want to lint. I don't think I can reasonably review this without just removing it entirely, and writing it from scratch and hoping I come up with a close enough final result.

That said, the diff makes sense and looks like it does what T-lang wants, I'm just not confident in the full behaviour of this lint now, the cases it's missing and the cases it's not missing.

I'm not sure my feelings should be blocking it, but I also don't want to pass on the hot potato for long enough until someone feels like it's not important to understand it fully.

**Comment by Urgau** [comment]
> and the exact things we want to lint

My understanding is that the lint should fire for what TC calls "sneaky inner impl".

What I understand from this is that the lint should warn on `impl` defs where syntactically none of "part" (idk the right word) of the `Self` and `Trait` types refer to something at the same nesting as the `impl` def it-self.

In particular it should not take into account the perfect type-system definition since T-lang doesn't "feel" like impls that would lint because of the 1-impl rule are "sneaky".

I think one could say that an `impl` def is sneaky if is hidden (that is nested inside other items) and that none of the syntactic part[^1] of the `Self` and `Trait` types refer to something at the same nesting as the `impl` definition it-self.

*I think it's also clear from [unanimous decision](https://github.com/rust-lang/rust/issues/126768#issuecomment-2192634762) from T-lang that they were not looking at a perfect non local `impl` defs detection, which the above rule is nowhere near.*

@traviscross could you confirm or deny what I just said.

[^1]: in the compiler those are `{ast,hir}::Path`, idk if we have a word or acronym at the syntax level

**Comment by BoxyUwU** [comment]
I can review this if you want, I only reassigned because I figured waffle might want to review this since it was on the original PR :woman_shrugging:

**Comment by workingjubilee** [comment]
Let's go!

r? @BoxyUwU

**Comment by traviscross** [comment]
@Urgau: That sounds right to me, but I'd probably point more normatively to my earlier comments if there were any questions here.

**Comment by BoxyUwU** [comment]
Sorry for taking so long to get to this, I've been looking over this the past few days and will hopefully be done soon

**Comment by BoxyUwU** [comment]
I think the `did_has_local_parent`/`path_has_local_parent` logic is written confusingly which is why it took a while to wrap my head around how this file worked. That seems to be pre-existing so I'm fine to land this anyway.

@bors r+

**Comment by Urgau** [comment]
Same as last times (https://github.com/rust-lang/rust/pull/120393#issuecomment-1971787475, https://github.com/rust-lang/rust/pull/125722#issuecomment-2169496199), the version of diesel used by `rustc-perf` is too old and use named `const AA` items instead of un-named `const _` items, making the lint trigger many times (more than 150 times last time). We ruled in the past that given the regression where located in one benchmark only, and only a pathological case that the regressions where fine. Marking as triaged.

@rustbot label: +perf-regression-triaged
cc https://github.com/rust-lang/rustc-perf/issues/1819

**Pull request merged**